utils/zerovec/src/ule/mod.rs

```diff
@@ -14,24 +14,34 @@ pub use chars::CharULE;
 pub use plain::PlainOldULE;
 
 use core::{mem, slice};
+use displaydoc::Display;
+
+/// Error type when parsing a byte slice into a ULE slice.
+#[derive(Debug, Display)]
+pub enum UleError<E> {
+    /// The byte slice is invalid for the given ULE type.
+    #[displaydoc("Invalid byte sequence")]
+    InvalidBytes(E),
+
+    /// The byte slice did not have the correct length for the given ULE type.
+    #[displaydoc("Invalid number of bytes")]
+    InvalidLength,
+}
 
 /// Fixed-width, byte-aligned data that can be cast to and from a little-endian byte slice.
 ///
 /// "ULE" stands for "Unaligned little-endian"
 ///
 /// # Safety
 ///
-/// There must be no padding bytes involved in this type: [`Self::as_byte_slice()`] *must* return
-/// a slice of initialized bytes provided that `Self` is initialized.
+/// Safety checklist for `ULE`:
 ///
-/// [`ULE::validate_byte_slice()`] *must* be implemented to validate a byte slice and return error
-/// for the same slices as [`ULE::parse_byte_slice()`].
-///
-/// [`ULE::from_bytes_unchecked()`] *must* be implemented to return the same result as [`ULE::parse_byte_slice()`],
-/// and both should return slices to the same region of memory.
-///
-/// [`ULE::as_byte_slice()`] should return a slice that is the in-memory representation of `Self`,
-/// i.e. it should be just a pointer cast, and `mem::size_of_val(self) == mem::size_of_val(self.as_byte_slice())`=
+/// 1. The type *must not* include any uninitialized or padding bytes.
+/// 2. The impl of [`ULE::validate_byte_slice()`] *must* return an error if the given byte slice
+///    would not represent a valid slice of this type.
+/// 3. All other methods *must* be left with their default impl, or else implemented according to
+///    their respective safety guidelines.
+/// 4. Acknowledge the following note about the equality invariant.
```

Comment:
Praise: I am glad we are adding this checklist; I was looking for something along these lines while writing my ULE impl.

Question: It's implicit in point 2, but is it worth calling out the `length % sizeof(X)` check explicitly, either as an additional bullet point or as an additional sentence in point 2?

Reply:
Added a new checklist item.